### ISSUE TITLE
Consul Namer and health-check status

### ISSUE THREAD

**Issue by Ashald** [BuoyantIO/linkerd#589]
Right now Consul's `CatalogNamer` uses `catalog` API and just fetches info about all known endpoints for the service regardless of their status. Although it plays well with the `linekrd` in general there 2 edge-cases that are not covered:

# Problem
1) In Consul there is so called `maintenance-mode`. It can be enabled for a particular service or a whole node. When enabled it adds a failing health-check to service (or each service if applied to a whole node). The purpose of this mode is to manually prohibit routing of requests to services registered within a node.
2) When `namerd` used not only as a backend for `linkerd` but also as a stand-alone resolver it doesn't have a feedback from `linkerd` and therefore HTTP API endpoints such as `delegate`, `addr`, `bind` and `resolve` don't know which of service endpoints are healthy and which a not.

# Possible Solutions
1) An additional Consul namer that would rely on `health` API https://www.consul.io/docs/agent/http/health.html (as it's done in `finagle-consul`) rather than `catalog` API and will be filtering out unhealthy services (proposed by @olix0r )
2) Switch `CatalogNamer` to `health` API and add an option to return all services or only healthy ones (can be done as easy as adding a `?passing` query parameter to the HTTP request)

I'd be happy to submit a PR implementing the feature in any form. It'd be great to hear other opinions so we can decide on best approach to handle the problem.

**Comment by JustinVenus** [comment]
I hit this same issue this past Friday, but didn't have time to look into it until this morning.  I think the easiest solution is to add '?passing' as suggested in the initial report to the code in [consul/src/main/scala/io/buoyant/consul/v1/CatalogApi.scala](
https://github.com/BuoyantIO/linkerd/blob/master/consul/src/main/scala/io/buoyant/consul/v1/CatalogApi.scala#L53).

**Comment by Ashald** [comment]
The `?passing` parameter is not supported by Catalog HTTP API. :(

**Comment by adleong** [comment]
I think implementing a second consul namer that uses the health API is a good starting place.  That way we can evaluate it alongside the existing consul namer that uses the catalog API.  If we find that the namer that uses the health API is more useful, we can potentially remove the catalog namer at a later date.

**Comment by JustinVenus** [comment]
FWIW this seems inter changeable in response as long as the additional json can easily ignored.  

Please ignore duplicate entry as that is a different issue on my side.

The node that is down in my environment is mesos-slave05of2.  Here is some output.

## Health Endpoint (Passing Only)
http://consul01dv2:8500/v1/health/service/cadence?dc=zerozero&tag=devel&passing
```javascript
[
{
Node: {
Node: "mesos-slave01of2",
Address: "10.150.150.227",
TaggedAddresses: null,
CreateIndex: 4477041,
ModifyIndex: 4596032
},
Service: {
ID: "v2.sd-bright-devel-cadence-member_0000000015",
Service: "cadence",
Tags: [
"devel",
"shard:0",
"user:bright"
],
Address: "",
Port: 31204,
EnableTagOverride: false,
CreateIndex: 4478435,
ModifyIndex: 4478446
},
Checks: [
{
Node: "mesos-slave01of2",
CheckID: "serfHealth",
Name: "Serf Health Status",
Status: "passing",
Notes: "",
Output: "Agent alive and reachable",
ServiceID: "",
ServiceName: "",
CreateIndex: 4477041,
ModifyIndex: 4477041
},
{
Node: "mesos-slave01of2",
CheckID: "service:v2.sd-bright-devel-cadence-member_0000000015",
Name: "Service 'cadence' check",
Status: "passing",
Notes: "",
Output: "TCP connect localhost:31204: Success",
ServiceID: "v2.sd-bright-devel-cadence-member_0000000015",
ServiceName: "cadence",
CreateIndex: 4478436,
ModifyIndex: 4478446
}
]
},
{
Node: {
Node: "mesos-slave01of2",
Address: "10.150.150.227",
TaggedAddresses: null,
CreateIndex: 4477041,
ModifyIndex: 4596032
},
Service: {
ID: "v2.sd-bright-devel-cadence-member_0000000016",
Service: "cadence",
Tags: [
"devel",
"shard:0",
"user:bright"
],
Address: "",
Port: 31204,
EnableTagOverride: false,
CreateIndex: 4482833,
ModifyIndex: 4482864
},
Checks: [
{
Node: "mesos-slave01of2",
CheckID: "serfHealth",
Name: "Serf Health Status",
Status: "passing",
Notes: "",
Output: "Agent alive and reachable",
ServiceID: "",
ServiceName: "",
CreateIndex: 4477041,
ModifyIndex: 4477041
},
{
Node: "mesos-slave01of2",
CheckID: "service:v2.sd-bright-devel-cadence-member_0000000016",
Name: "Service 'cadence' check",
Status: "passing",
Notes: "",
Output: "TCP connect localhost:31204: Success",
ServiceID: "v2.sd-bright-devel-cadence-member_0000000016",
ServiceName: "cadence",
CreateIndex: 4482834,
ModifyIndex: 4482864
}
]
}
]
```

## Health Endpoint Everything
http://consul01dv2:8500/v1/health/service/cadence?dc=zerozero&tag=devel
```javascript
[
{
Node: {
Node: "mesos-slave01of2",
Address: "10.150.150.227",
TaggedAddresses: null,
CreateIndex: 4477041,
ModifyIndex: 4596032
},
Service: {
ID: "v2.sd-bright-devel-cadence-member_0000000015",
Service: "cadence",
Tags: [
"devel",
"shard:0",
"user:bright"
],
Address: "",
Port: 31204,
EnableTagOverride: false,
CreateIndex: 4478435,
ModifyIndex: 4478446
},
Checks: [
{
Node: "mesos-slave01of2",
CheckID: "serfHealth",
Name: "Serf Health Status",
Status: "passing",
Notes: "",
Output: "Agent alive and reachable",
ServiceID: "",
ServiceName: "",
CreateIndex: 4477041,
ModifyIndex: 4477041
},
{
Node: "mesos-slave01of2",
CheckID: "service:v2.sd-bright-devel-cadence-member_0000000015",
Name: "Service 'cadence' check",
Status: "passing",
Notes: "",
Output: "TCP connect localhost:31204: Success",
ServiceID: "v2.sd-bright-devel-cadence-member_0000000015",
ServiceName: "cadence",
CreateIndex: 4478436,
ModifyIndex: 4478446
}
]
},
{
Node: {
Node: "mesos-slave01of2",
Address: "10.150.150.227",
TaggedAddresses: null,
CreateIndex: 4477041,
ModifyIndex: 4596032
},
Service: {
ID: "v2.sd-bright-devel-cadence-member_0000000016",
Service: "cadence",
Tags: [
"devel",
"shard:0",
"user:bright"
],
Address: "",
Port: 31204,
EnableTagOverride: false,
CreateIndex: 4482833,
ModifyIndex: 4482864
},
Checks: [
{
Node: "mesos-slave01of2",
CheckID: "serfHealth",
Name: "Serf Health Status",
Status: "passing",
Notes: "",
Output: "Agent alive and reachable",
ServiceID: "",
ServiceName: "",
CreateIndex: 4477041,
ModifyIndex: 4477041
},
{
Node: "mesos-slave01of2",
CheckID: "service:v2.sd-bright-devel-cadence-member_0000000016",
Name: "Service 'cadence' check",
Status: "passing",
Notes: "",
Output: "TCP connect localhost:31204: Success",
ServiceID: "v2.sd-bright-devel-cadence-member_0000000016",
ServiceName: "cadence",
CreateIndex: 4482834,
ModifyIndex: 4482864
}
]
},
{
Node: {
Node: "mesos-slave05of2",
Address: "10.150.150.231",
TaggedAddresses: null,
CreateIndex: 4458480,
ModifyIndex: 4611589
},
Service: {
ID: "v2.sd-bright-devel-cadence-member_0000000014",
Service: "cadence",
Tags: [
"devel",
"shard:0",
"user:bright"
],
Address: "",
Port: 31308,
EnableTagOverride: false,
CreateIndex: 4478149,
ModifyIndex: 4478418
},
Checks: [
{
Node: "mesos-slave05of2",
CheckID: "serfHealth",
Name: "Serf Health Status",
Status: "passing",
Notes: "",
Output: "Agent alive and reachable",
ServiceID: "",
ServiceName: "",
CreateIndex: 4458480,
ModifyIndex: 4458480
},
{
Node: "mesos-slave05of2",
CheckID: "service:v2.sd-bright-devel-cadence-member_0000000014",
Name: "Service 'cadence' check",
Status: "critical",
Notes: "",
Output: "dial tcp [::1]:31308: getsockopt: connection refused",
ServiceID: "v2.sd-bright-devel-cadence-member_0000000014",
ServiceName: "cadence",
CreateIndex: 4478149,
ModifyIndex: 4478418
}
]
}
]
```

## Catalog Endpoint (current source of truth)
http://consul01dv2:8500/v1/catalog/service/cadence?dc=zerozero&tag=devel
```javascript
[
{
Node: "mesos-slave01of2",
Address: "10.150.150.227",
ServiceID: "v2.sd-bright-devel-cadence-member_0000000015",
ServiceName: "cadence",
ServiceTags: [
"devel",
"shard:0",
"user:bright"
],
ServiceAddress: "",
ServicePort: 31204,
ServiceEnableTagOverride: false,
CreateIndex: 4478435,
ModifyIndex: 4478446
},
{
Node: "mesos-slave01of2",
Address: "10.150.150.227",
ServiceID: "v2.sd-bright-devel-cadence-member_0000000016",
ServiceName: "cadence",
ServiceTags: [
"devel",
"shard:0",
"user:bright"
],
ServiceAddress: "",
ServicePort: 31204,
ServiceEnableTagOverride: false,
CreateIndex: 4482833,
ModifyIndex: 4482864
},
{
Node: "mesos-slave05of2",
Address: "10.150.150.231",
ServiceID: "v2.sd-bright-devel-cadence-member_0000000014",
ServiceName: "cadence",
ServiceTags: [
"devel",
"shard:0",
"user:bright"
],
ServiceAddress: "",
ServicePort: 31308,
ServiceEnableTagOverride: false,
CreateIndex: 4478149,
ModifyIndex: 4478418
}
]
```

**Comment by JustinVenus** [comment]
I have a pull request for this issue open https://github.com/BuoyantIO/linkerd/pull/595.

**Comment by Ashald** [comment]
@adleong I have plans to submit a PR [hopefully] somewhere next week implementing it as an extra namer. Want to bounce an idea for it with you first:
1) By default behaves same way as `CatalogNamer` but has an option to respect Consul's health status (by adding `?passing`)
2) By default filters out services in 'maintenance mode' but doesn't pay attention to other health checks _and_ has an option to return only healthy services (by adding `?passing`)
3) Just returns only healthy services all the time (always add `?passing`)
4) Accepts names/patterns for health check names/IDs that it should respect and ignore other
5) Something else?

Thanks!

**Comment by adleong** [comment]
@Ashald you may want to coordinate with @JustinVenus who has started on this in https://github.com/BuoyantIO/linkerd/pull/595

**Comment by olix0r** [comment]
I believe this was merged in #595.  Please feel free to reopen if there are remaining issues.
